aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -11,26 +11,47 @@ use crate::{
     VMStatus,
 };
 use move_binary_format::file_format_common::read_uleb128_as_u64;
-use move_core_types::{account_address::AccountAddress, value::MoveValue, vm_status::StatusCode};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::ModuleId,
+    value::MoveValue,
+    vm_status::StatusCode,
+};
 use move_vm_runtime::session::LoadedFunctionInstantiation;
-use move_vm_types::loaded_data::runtime_types::Type;
+use move_vm_types::{gas::GasMeter, loaded_data::runtime_types::Type};
 use once_cell::sync::Lazy;
 use std::{
     collections::BTreeMap,
     io::{Cursor, Read},
 };
+use move_vm_types::gas::UnmeteredGasMeter;
 
-// A map which contains the structs allowed as transaction input and the
-// validation function for those, if one was needed (None otherwise).
-// The validation function takes the serialized argument and returns
-// an error if the validation fails.
-type ValidateArg = fn(&[u8]) -> Result<(), VMStatus>;
+pub(crate) struct FunctionId {
+    module_id: ModuleId,
+    func_name: &'static str,
+}
 
-static ALLOWED_STRUCTS: Lazy<BTreeMap<String, Option<ValidateArg>>> = Lazy::new(|| {
-    [("0x1::string::String", Some(check_string as ValidateArg))]
-        .into_iter()
-        .map(|(s, validator)| (s.to_string(), validator))
-        .collect()
+static ALLOWED_STRUCTS: Lazy<BTreeMap<String, FunctionId>> = Lazy::new(|| {
+    [
+        ("0x1::string::String", FunctionId {
+            module_id: ModuleId::new(
+                AccountAddress::ONE,
+                Identifier::new("string").expect("cannot fail"),
+            ),
+            func_name: "utf8",
```

Comment:
Can we do this in a separate PR with a separate feature flag? Just to make sure we keep unrelated changes out of making Object<T> an input type

Reply:
Do you mean keep out how to deal with strings @movekevin ? This would not work good, because this PR unifies the way how strings and objects are dealt with. You actually like to show and test with this code that it works for both strings and objects.

Reply:
I refactored the code to be very close to the original. It should be clear now that there are no functional differences for the string case.